adapters/conversant/conversant.go

```diff
@@ -1,291 +1,177 @@
-	TagID       string   `json:"tag_id"`
-	Position    *int8    `json:"position"`
-	BidFloor    float64  `json:"bidfloor"`
-	Mobile      *int8    `json:"mobile"`
-	MIMEs       []string `json:"mimes"`
-	API         []int8   `json:"api"`
-	Protocols   []int8   `json:"protocols"`
-	MaxDuration *int64   `json:"maxduration"`
-}
-
-func (a *ConversantAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder *pbs.PBSBidder) (pbs.PBSBidSlice, error) {
-	mediaTypes := []pbs.MediaType{pbs.MEDIA_TYPE_BANNER, pbs.MEDIA_TYPE_VIDEO}
-	cnvrReq, err := adapters.MakeOpenRTBGeneric(req, bidder, a.Name(), mediaTypes)
-
-	if err != nil {
-		return nil, err
+func (c ConversantAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("No impressions in bid request"),
+		}}
 	}
 
-	// Create a map of impression objects for both request creation
-	// and response parsing.
-
-	impMap := make(map[string]*openrtb.Imp, len(cnvrReq.Imp))
-	for idx := range cnvrReq.Imp {
-		impMap[cnvrReq.Imp[idx].ID] = &cnvrReq.Imp[idx]
-	}
-
-	// Fill in additional info from custom params
-
-	for _, unit := range bidder.AdUnits {
-		var params conversantParams
-
-		imp := impMap[unit.Code]
-		if imp == nil {
-			// Skip ad units that do not have corresponding impressions.
-			continue
+	for i := 0; i < len(request.Imp); i++ {
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(request.Imp[i].Ext, &bidderExt); err != nil {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Impression[%d] missing ext object", i),
+			}}
 		}
 
-		err := json.Unmarshal(unit.Params, &params)
-		if err != nil {
-			return nil, &errortypes.BadInput{
-				Message: err.Error(),
-			}
+		var cnvrExt openrtb_ext.ExtImpConversant
+		if err := json.Unmarshal(bidderExt.Bidder, &cnvrExt); err != nil {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Impression[%d] missing ext.bidder object", i),
+			}}
 		}
 
-		// Fill in additional Site info
-		if params.SiteID != "" {
-			if cnvrReq.Site != nil {
-				cnvrReq.Site.ID = params.SiteID
-			}
-			if cnvrReq.App != nil {
-				cnvrReq.App.ID = params.SiteID
-			}
+		if cnvrExt.SiteID == "" {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Impression[%d] requires ext.bidder.site_id", i),
+			}}
 		}
 
-		if params.Mobile != nil && !(cnvrReq.Site == nil) {
-			cnvrReq.Site.Mobile = *params.Mobile
-		}
-
-		// Fill in additional impression info
-
-		imp.DisplayManager = "prebid-s2s"
-		imp.DisplayManagerVer = "1.0.1"
-		imp.BidFloor = params.BidFloor
-		imp.TagID = params.TagID
-
-		var position *openrtb.AdPosition
-		if params.Position != nil {
-			position = openrtb.AdPosition(*params.Position).Ptr()
+		if i == 0 {
+			if request.Site != nil {
+				request.Site.ID = cnvrExt.SiteID
```

Comment:
`request.Site.ID` and `request.App.ID` are shared memory and should not be directly changed. Please make a copy first, something like:

```
siteCopy := *request.Site
request.Site = &siteCopy
request.Site.ID = cnvrExt.SiteID
```

... as well as for App.ID and a few other places where I've commented. This is the root cause of the race condition test failures reported by Travis (yay! they work!).